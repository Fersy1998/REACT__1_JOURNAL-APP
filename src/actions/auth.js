import { types } from "../types/types"
import Swal from 'sweetalert2'
import React from 'react';

import {firebase, googleAuthProvider} from '../firebase/firebase-config';
import { finishLoading, startLoading } from "./ui";
import { Navigate, useNavigate } from "react-router-dom";
import { noteLogout } from "./notes";


export const startLoginEmailPassword=(email, password)=>{
    
    return (dispatch)=>{
        dispatch(startLoading());
        return firebase.auth().signInWithEmailAndPassword(email, password)
        .then(async ({user})=>{
            //console.log(res);
            dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
        }).catch((e)=>{
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error')
        })
       
    }
/*
    return (dispatch)=>{
        setTimeout(()=>{
            dispatch(login(email, password));
        
        }, 3500)
    }*/
}

export const startRegisterWithEmalPasswordName=(email, password, name)=>{
    return (dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email, password)
        .then(async ({user})=>{   
            //console.log(user);
            await user.updateProfile({displayName:name});
            dispatch(login(user.uid, user.displayName));
        
    }).catch((e)=>{
        //dispatch(finishLoading());
        Swal.fire('Error', e.message, 'error')
    })
    }
}
export const startGoogleLogin=()=>{
    return (dispatch)=>{
        firebase.auth().signInWithPopup(googleAuthProvider.setCustomParameters({prompt: "select_account"}))
        /*.then(userCred=>{
            console.log(userCred);
        })*/
      .then(({user})=>{   
            console.log(user);
            dispatch(login(user.uid, user.displayName));
        }
        )
    }
}

export const login=(uid, displayname)=>{
    
    
    return {
        type:types.login,
        payload:{
            uid,
            displayname
        }
    }
}

export const startLogOut=()=>{
    return async (dispatch)=>{
        await firebase.auth().signOut()
        dispatch(logOut());
        dispatch(noteLogout());
    }

}
export const logOut=()=>{
    return {
        type:types.logout
    } 
}
