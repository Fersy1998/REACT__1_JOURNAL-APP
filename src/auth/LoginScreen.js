import React, { useEffect, useState } from 'react'
import { Link, Navigate, useNavigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../hooks/useForm'
import {setError, removeError} from '../actions/ui';
import { startGoogleLogin, startLoginEmailPassword } from '../actions/auth'
import validator from 'validator';
export const LoginScreen = () => {
    
    const navigate=useNavigate();
    const dispatch=useDispatch()
    const {msgError, loading}=useSelector(state=>state.ui);
    const user=useSelector(state=>state.auth)
    
    const [formValues, handleInputChange]=useForm({
        email:'fersy@gmail.com',
        password:'1hwdhid'
    })
    const [userActive, setActiveUser] = useState(user);
    useEffect(() => {
        if(user?.uid){
            setActiveUser(true);
            navigate('/auth/journal', {
                replace:true
              }) ;
          }else{
            setActiveUser(false);
          }
    }, [setActiveUser])
    

    
    
    const {email, password}=formValues;
    const handleLogin=(e)=>{
        e.preventDefault();
        //dispatch(login(892183127, 'Fersy'));
        if (isFormValid()){
            dispatch(startLoginEmailPassword(email, password));
            navigate('/', {
                replace:true
              }) ;
        }
    }
    const handleGoogleLogin=()=>{
        if(isFormValid()){
            dispatch(startGoogleLogin());
            
        }
        
    }
    const isFormValid=()=>{
        if(!validator.isEmail(email)){
          dispatch(setError('Invalid email'));
          return false;
        }else if(password.length<6 ){
          dispatch(setError('password should have at least 6 characters and match'));
          return false
        }
        dispatch(removeError());
        return true
      
      }
  return (

    <div className='animate__animated animate__headShake'>
        <h3 className='auth__title'>
            Login
        </h3>
        <form onSubmit={handleLogin}>
            {
            msgError &&
            (<div className='auth__alert-error'>{msgError}</div>)
            }
            <input type='text' placeholder='email' onChange= {handleInputChange} name='email' className='auth__input' value={email}/>
            <input type='password' placeholder='password' onChange= {handleInputChange} name='password' className='auth__input' value={password}/>
            <button type='submit' className='btn btn-primary btn-block' disabled={loading}>Login</button><hr />
            <div className='auth__social-network'>
                <p>Login with social Network</p>
                <div className="google-btn" onClick={handleGoogleLogin}>
                    <div className="google-icon-wrapper">
                        <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                    </div>
                    <p className="btn-text">
                        <b>Sign in with google</b>
                    </p>
                </div>
            </div>
            <Link to='/auth/register' className='link'>
                Create New Acount
            </Link>
        </form>
        
    </div>

  )
}
