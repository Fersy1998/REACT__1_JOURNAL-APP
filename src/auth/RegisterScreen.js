import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../hooks/useForm'
import validator from 'validator';
import {setError, removeError} from '../actions/ui';
import { startRegisterWithEmalPasswordName } from '../actions/auth';
export const RegisterScreen = () => {

  const [formValues, handleInputChange]=useForm({name:'Fersy', email:'fersy@yahoo.com', password:'123', password2:'123'});
  const dispatch=useDispatch();
  
  //const state=useSelector(state=>state.ui); extrae el estado de Redux del navegador
  const {msgError}=useSelector(state=>state.ui);
  const {name, email, password, password2}=formValues;
  
  
  const handleRegister=(e)=>{
    e.preventDefault();
    
    if(isFormValid()){
      dispatch(startRegisterWithEmalPasswordName(email, password, name));
      //console.log(formValues);
    }
  
  }
  const isFormValid=()=>{
    if(name.trim().length===0){
      dispatch(setError('name is required'));
      return false;
    }else if(!validator.isEmail(email)){
      dispatch(setError('Invalid email'));
      return false;
    }else if(password!==password2 || password.length<6 ){
      dispatch(setError('password should have at least 6 characters and match'));
      return false
    }
    dispatch(removeError());
    return true
  
  }
  return (
    
    <div className='animate__animated animate__headShake'>
        <h3 className='auth__title' >
            Register
        </h3>
        <form onSubmit={handleRegister}>
            {
            msgError &&
            (<div className='auth__alert-error'>{msgError}</div>)
            }
            <input type='text' placeholder='name' name='name' className='auth__input'value={name} onChange={handleInputChange}/>
            <input type='text' placeholder='email' name='email' className='auth__input'value={email} onChange={handleInputChange}/>
            <input type='password' placeholder='password' name='password' className='auth__input' value={password} onChange={handleInputChange}/>
            <input type='password' placeholder='confirm password' name='password2' className='auth__input' value={password2} onChange={handleInputChange}/>
            <button type='submit' className='btn btn-primary btn-block'>Login</button><hr />
            <Link to='auth/login' className='link'>
                Already register?
            </Link>
        </form>
        
    </div>
 
  )
}
