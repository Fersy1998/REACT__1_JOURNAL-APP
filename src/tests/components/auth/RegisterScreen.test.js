import React from 'react';
import '@testing-library/jest-dom';


import { mount } from "enzyme";
import { RegisterScreen } from "../../../auth/RegisterScreen";

import { Provider } from "react-redux";
import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { MemoryRouter } from "react-router-dom";


jest.mock("../../../actions/auth", ()=>({ 
    startGoogleLogin: jest.fn() ,
    startLoginEmailPassword:jest.fn(),
    
}))
jest.setTimeout(9000);

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState={
    auth: {},
    ui: {
      loading: false,
      msgError: 'Invalid email'
    },
    notes: {
      notes: [],
      active: null
    }
  }
let store=mockStore(initState);

//store.dispatch=jest.fn()
//Mmemory router finge la rutas
const wrapper=mount(
    <Provider store={store}>
        <MemoryRouter >
            <RegisterScreen />
        </MemoryRouter>
    </Provider> );

describe('Pruebas en registerScreen.js', ()=>{
    beforeEach(()=>{
        store=mockStore(initState);
        jest.clearAllMocks();
    })
    test('should match the snapshot', () => { 
        expect(wrapper).toMatchSnapshot();
    })
    test('should do the respective action', () => {
        const emailField=wrapper.find('input[name="email"]');
        const nameField=wrapper.find('input[name="email"]');
        nameField.simulate('change', {
            target:{
                value:'Fersy',
                name:'name'
            }
        })
        console.log(emailField);
        emailField.simulate('change', {
            target:{
                value:'',
                name:'email'
            }
        })
        wrapper.find('form').prop('onSubmit')({preventDefault(){}});
        const actions=store.getActions();
        console.log(actions);
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe('Invalid email');
    })
    

})