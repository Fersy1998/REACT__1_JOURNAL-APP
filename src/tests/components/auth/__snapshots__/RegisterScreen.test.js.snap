// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en registerScreen.js should match the snapshot 1`] = `
<div
  className="animate__animated animate__headShake"
>
  <h3
    className="auth__title"
  >
    Register
  </h3>
  <form
    onSubmit={[Function]}
  >
    <div
      className="auth__alert-error"
    >
      Invalid email
    </div>
    <input
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="name"
      type="text"
      value="Fersy"
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="email"
      type="text"
      value="fersy@yahoo.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value="123"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="confirm password"
      type="password"
      value="123"
    />
    <button
      className="btn btn-primary btn-block"
      type="submit"
    >
      Login
    </button>
    <hr />
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>
</div>
`;
